### PR TITLE
Detach sample and log_prob from neural net

### PR DESCRIPTION
**Pull request by michaeldeistler**
This PR provides methods that allow to `sample()` and `log_prob()` from a MoG when already knowing the precision and means. It's a simple refactor, but it will save us a lot of code repetition [here](https://github.com/mackelab/sbi/pull/458).

### CODE REVIEW

File: pyknos/mdn/mdn.py [hunk 1]
```diff
@@ -160,7 +160,35 @@ def log_prob(self, inputs: Tensor, context=Optional[Tensor]) -> Tensor:
...
+        logits: Tensor,
+        means: Tensor,
+        precisions: Tensor,
+        sumlogdiag: Tensor,
+    ) -> Tensor:
+        """
+        Return the log-probability of `inputs` under a MoG with specified parameters.
+
+        Unlike the `log_prob()` method, this method is fully detached from the neural
+        network and is a `staticmethod`.
```
Comment (janfb) [inline pyknos/mdn/mdn.py]: would remove the "and is a static method", this is clear from the code, no? 
Instead I would add why this is useful, e.g., add sth like "can be used independent of the neural net in case the MoG parameters are already known"?

File: pyknos/mdn/mdn.py [hunk 2]
```diff
@@ -195,6 +223,30 @@ def sample(self, num_samples: int, context: Tensor) -> Tensor:
...
+        return self.sample_mog(num_samples, logits, means, precision_factors)
+
+    @staticmethod
+    def sample_mog(
+        num_samples: int, logits: Tensor, means: Tensor, precision_factors: Tensor
+    ) -> Tensor:
+        """
+        Return samples of a MoG with specified parameters.
+
+        Unlike the `sample()` method, this method is fully detached from the neural
```
Comment (janfb) [inline pyknos/mdn/mdn.py:236]: same addition as for the `log_prob_mog` docstring.

**Review by janfb — APPROVED** [review]
Looks good, thanks! 
Can be merged once the two comments are addressed.

**Comment by michaeldeistler** [comment]
Totally agree with both comments, thanks!
